cylc/flow/suite_status.py

```diff
@@ -27,19 +29,115 @@
 KEY_UPDATE_TIME = "update-time"
 KEY_VERSION = "version"
 
-# Suite status strings.
-SUITE_STATUS_HELD = "held"
-SUITE_STATUS_RUNNING = "running"
-SUITE_STATUS_STOPPING = "stopping"
+# Status message strings
 SUITE_STATUS_RUNNING_TO_STOP = "running to stop at %s"
 SUITE_STATUS_RUNNING_TO_HOLD = "running to hold at %s"
 
-# Pseudo status strings for use by suite monitors.
-#   Use before attempting to determine status:
-SUITE_STATUS_NOT_CONNECTED = "not connected"
-#   Use prior to first status update:
-SUITE_STATUS_CONNECTED = "connected"
-SUITE_STATUS_INITIALISING = "initialising"
-#   Use when the suite is not running:
-SUITE_STATUS_STOPPED = "stopped"
-SUITE_STATUS_STOPPED_WITH = "stopped with '%s'"
+# TODO - the suite status should be a property of the Scheduler instance
+#        rather than derived from scheduler state?
+
+
+class SuiteStatus(Enum):
+    """The possible statuses of a suite."""
+
+    HELD = "held"
+    """Suite will not submit any new jobs."""
+
+    RUNNING = "running"
+    """Suite is running as normal."""
+
+    STOPPING = "stopping"
+    """Suite is in the process of shutting down."""
+
+    STOPPED = "stopped"
+    """Suite is not running."""
+
+
+class StopMode(Enum):
+    """The possible modes of a suite shutdown"""
+
+    AUTO = 'AUTOMATIC'
+    """Suite has reached a state where it can automatically stop"""
+
+    AUTO_ON_TASK_FAILURE = 'AUTOMATIC(ON-TASK-FAILURE)'
+    """A task has failed and ``abort if any task fails = True``."""
+
+    REQUEST_CLEAN = 'REQUEST(CLEAN)'
+    """External shutdown request, will wait for active jobs to complete."""
+
+    REQUEST_NOW = 'REQUEST(NOW)'
+    """External shutdown request, will wait for event handlers to complete."""
+
+    REQUEST_NOW_NOW = 'REQUEST(NOW-NOW)'
+    """External immediate shutdown request."""
+
+    def describe(self):
+        """Return a user-friendly description of this state."""
+        if self == self.AUTO:
+            return 'suite has completed'
+        if self == self.AUTO_ON_TASK_FAILURE:
+            return 'a task has finished'
+        if self == self.REQUEST_CLEAN:
+            return 'waiting for active jobs to complete'
+        if self == self.REQUEST_NOW:
+            return 'waiting for event handlers to complete'
+        if self == self.REQUEST_NOW_NOW:
+            return 'immediate shutdown'
```

Comment:
Can these values be the same as the docstrings?

Reply:
Sadly no.

Python did consider attribute docstrings (see the doomed [PEP224](https://www.python.org/dev/peps/pep-0224/)) but at present attributes are not objects which can have a `__doc__` value. Systems like Sphinx have their own parsers for auto-documentation purposes.

Reply:
If you were asking why these strings are different to the docstrings above, the reason is:

* The `describe()` strings ate for sending out on the network (for the GUI)
* Whereas the docstrings are for auto-documentation (if desired)